File: claude_exp/xv4/shared/stats.py
import threading
import time
from collections import deque


class Stats:
    WINDOWS = [30, 60, 180, 1800]

    def __init__(self, yellow_threshold_seconds=None):
        self._lock = threading.Lock()
        self._sent_times = deque()
        self._recv_times = deque()
        self._sent_total = 0
        self._recv_total = 0
        self._last_recv_time = None
        self._connections = {}
        self._gauges = {}
        self._yellow_threshold_seconds = yellow_threshold_seconds

    def set_connection(self, name: str, connected: bool):
        with self._lock:
            self._connections[name] = connected

    def set_gauge(self, name: str, value):
        with self._lock:
            self._gauges[name] = value

    def record_sent(self):
        with self._lock:
            now = time.monotonic()
            self._sent_times.append(now)
            self._sent_total += 1

    def record_recv(self):
        with self._lock:
            now = time.monotonic()
            self._recv_times.append(now)
            self._recv_total += 1
            self._last_recv_time = now

    def _count_since(self, times_deque, cutoff):
        count = 0
        for t in reversed(times_deque):
            if t >= cutoff:
                count += 1
            else:
                break
        return count

    def snapshot(self) -> dict:
        with self._lock:
            now = time.monotonic()

            # Prune oldest entries beyond max window
            max_window = max(self.WINDOWS)
            cutoff_max = now - max_window
            while self._sent_times and self._sent_times[0] < cutoff_max:
                self._sent_times.popleft()
            while self._recv_times and self._recv_times[0] < cutoff_max:
                self._recv_times.popleft()

            result = {
                "sent_total": self._sent_total,
                "recv_total": self._recv_total,
            }

            for w in self.WINDOWS:
                cutoff = now - w
                result[f"sent_{w}s"] = self._count_since(self._sent_times, cutoff)
                result[f"recv_{w}s"] = self._count_since(self._recv_times, cutoff)

            if self._last_recv_time is not None:
                elapsed = now - self._last_recv_time
                result["seconds_since_last_recv"] = round(elapsed, 1)
                import datetime
                wall_offset = time.time() - time.monotonic()
                wall_last = self._last_recv_time + wall_offset
                result["last_recv_datetime"] = datetime.datetime.fromtimestamp(
                    wall_last
                ).strftime("%H:%M:%S")
            else:
                result["seconds_since_last_recv"] = None
                result["last_recv_datetime"] = None

            if self._yellow_threshold_seconds is not None:
                result["yellow_threshold_seconds"] = self._yellow_threshold_seconds

            if self._connections:
                result["connections"] = dict(self._connections)

            if self._gauges:
                result["gauges"] = dict(self._gauges)

            return result

File: claude_exp/routers/crypto_host/src/shared/stats.cpp
#include "shared/stats.h"

#include <array>
#include <chrono>
#include <cmath>
#include <ctime>

namespace xv6::shared {

namespace {

constexpr std::array<int, 4> kWindows = {30, 60, 180, 1800};
constexpr int kMaxWindowSeconds = 1800;

int64_t now_ms() {
    return std::chrono::duration_cast<std::chrono::milliseconds>(
               std::chrono::system_clock::now().time_since_epoch())
        .count();
}

}  // namespace

Stats::Stats(std::optional<int> yellow_threshold_seconds)
    : yellow_threshold_seconds_(yellow_threshold_seconds) {}

void Stats::set_connection(const std::string& name, bool connected) {
    std::lock_guard<std::mutex> lock(mutex_);
    connections_[name] = connected;
}

void Stats::set_gauge(const std::string& name, nlohmann::json value) {
    std::lock_guard<std::mutex> lock(mutex_);
    gauges_[name] = std::move(value);
}

void Stats::record_sent() {
    std::lock_guard<std::mutex> lock(mutex_);
    int64_t t = now_ms();
    sent_ts_.push_back(t);
    ++sent_total_;
    trim(sent_ts_, t);
}

void Stats::record_recv() {
    std::lock_guard<std::mutex> lock(mutex_);
    int64_t t = now_ms();
    recv_ts_.push_back(t);
    ++recv_total_;
    last_recv_ms_ = t;
    trim(recv_ts_, t);
}

void Stats::trim(std::deque<int64_t>& dq, int64_t now_ms_val) const {
    int64_t cutoff = now_ms_val - static_cast<int64_t>(kMaxWindowSeconds) * 1000;
    while (!dq.empty() && dq.front() < cutoff) {
        dq.pop_front();
    }
}

int Stats::count_within(const std::deque<int64_t>& dq, int64_t now_ms_val, int window_seconds) const {
    int64_t cutoff = now_ms_val - static_cast<int64_t>(window_seconds) * 1000;
    int count = 0;
    for (auto it = dq.rbegin(); it != dq.rend(); ++it) {
        if (*it >= cutoff) {
            ++count;
        } else {
            break;
        }
    }
    return count;
}

nlohmann::json Stats::snapshot() const {
    std::lock_guard<std::mutex> lock(mutex_);
    int64_t t = now_ms();

    nlohmann::json out;
    out["sent_total"] = sent_total_;
    out["recv_total"] = recv_total_;
    for (int w : kWindows) {
        out["sent_" + std::to_string(w) + "s"] = count_within(sent_ts_, t, w);
        out["recv_" + std::to_string(w) + "s"] = count_within(recv_ts_, t, w);
    }

    if (last_recv_ms_.has_value()) {
        double secs = static_cast<double>(t - *last_recv_ms_) / 1000.0;
        secs = std::round(secs * 10.0) / 10.0;
        out["seconds_since_last_recv"] = secs;

        std::time_t tt = static_cast<std::time_t>(*last_recv_ms_ / 1000);
        std::tm tm_buf{};
        localtime_r(&tt, &tm_buf);
        char buf[16];
        std::strftime(buf, sizeof(buf), "%H:%M:%S", &tm_buf);
        out["last_recv_datetime"] = std::string(buf);
    } else {
        out["seconds_since_last_recv"] = nullptr;
        out["last_recv_datetime"] = nullptr;
    }

    if (yellow_threshold_seconds_.has_value()) {
        out["yellow_threshold_seconds"] = *yellow_threshold_seconds_;
    }
    if (!connections_.empty()) {
        nlohmann::json conns = nlohmann::json::object();
        for (const auto& [k, v] : connections_) {
            conns[k] = v;
        }
        out["connections"] = std::move(conns);
    }
    if (!gauges_.empty()) {
        nlohmann::json g = nlohmann::json::object();
        for (const auto& [k, v] : gauges_) {
            g[k] = v;
        }
        out["gauges"] = std::move(g);
    }
    return out;
}

}  // namespace xv6::shared
